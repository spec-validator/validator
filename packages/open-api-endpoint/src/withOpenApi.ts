import { _, ServerConfig, getServerConfigs } from '@spec-validator/rest-api-server'
import { Json } from '@spec-validator/utils/Json'
import { segmentField as $, constantField, stringField, wildcardObjectField } from '@spec-validator/validator/fields'
import genOpenApi, { DEFAULT_INFO, Info, WithInfo } from './genOpenApi'

const getUI = (info: Info) => `
  <html>
  <head>
    <title>${info.title} : ${info.version}</title>

    <link rel="stylesheet" type="text/css" href="https://unpkg.com/swagger-ui-dist@3.38.0/swagger-ui.css">
  <body>

    <div id="swagger-ui"></div>

    <script src="https://unpkg.com/swagger-ui-dist@3/swagger-ui-bundle.js" charset="UTF-8"></script>
    <script src="https://unpkg.com/swagger-ui-dist@3/swagger-ui-standalone-preset.js" charset="UTF-8"></script>

    <script>
    window.onload = function() {
      const ui = SwaggerUIBundle({
        url: window.location.href.split('?')[0].replace(/-ui$/, ''),
        dom_id: '#swagger-ui',
        presets: [
          SwaggerUIBundle.presets.apis,
          SwaggerUIStandalonePreset
        ]
      })

      window.ui = ui
    }
    </script>

  </body>
  </html>
`

export default <C extends Partial<ServerConfig> & WithInfo>(config: C, schemaRoot = '/open-api'): ServerConfig => {
  const merged = getServerConfigs(config)
  const routes = [
    ...merged.routes,
    _.GET($._(schemaRoot)).spec(
      {
        response: {
          body: wildcardObjectField(),
        },
      },
    ).handler(
      async () => ({
        body: genOpenApi(merged) as unknown as Record<string, Json>,
      })
    ),
    _.GET($._(schemaRoot)._('-ui')).spec(
      {
        response: {
          body: stringField(),
          headers: {
            'content-type': constantField('text/html'),
          },
        },
      },
    ).handler(
      async () => ({
        body: getUI(config.info || DEFAULT_INFO),
        headers: {
          'content-type': 'text/html' as const,
        },
      })
    ),
  ]

  return {
    ...merged,
    routes,
  }
}
