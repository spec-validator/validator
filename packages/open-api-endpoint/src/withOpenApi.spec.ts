import withOpenApi from './withOpenApi'

const fullRoute = withOpenApi({}, '/schema')

test('routing', () => {
  expect(fullRoute).toMatchSnapshot()
})

test('schema root handler', async () => {
  expect(await fullRoute.routes[0].handler(undefined as any)).toMatchSnapshot()
})

test('schema root handler ui', async () => {
  expect(await fullRoute.routes[1].handler(undefined as any)).toMatchSnapshot()
})
