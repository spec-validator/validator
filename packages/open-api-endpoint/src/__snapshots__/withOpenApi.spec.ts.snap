// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`routing 1`] = `
Object {
  "appErrorStatusCode": 500,
  "baseUrl": "http://localhost:8000",
  "encoding": "utf-8",
  "frameworkErrorStatusCode": 503,
  "reportError": [Function],
  "routes": Array [
    Object {
      "handler": [Function],
      "request": Object {
        "method": Object {
          "constant": "GET",
          "getStringField": [Function],
          "regex": /GET/,
          "serialize": [Function],
          "type": "@spec-validator/validator/fields/constantField",
          "validate": [Function],
        },
        "pathParams": SegmentField {
          "field": undefined,
          "key": "/open-api",
          "parent": SegmentField {
            "field": undefined,
            "key": "",
            "parent": undefined,
            "type": "@spec-validator/validator/fields/segmentField",
          },
          "type": "@spec-validator/validator/fields/segmentField",
        },
      },
      "response": Object {
        "body": Object {
          "serialize": [Function],
          "type": "@spec-validator/validator/fields/wildcardObjectField",
          "validate": [Function],
        },
        "statusCode": Object {
          "constant": 200,
          "getStringField": [Function],
          "regex": /200/,
          "serialize": [Function],
          "type": "@spec-validator/validator/fields/constantField",
          "validate": [Function],
        },
      },
    },
    Object {
      "handler": [Function],
      "request": Object {
        "method": Object {
          "constant": "GET",
          "getStringField": [Function],
          "regex": /GET/,
          "serialize": [Function],
          "type": "@spec-validator/validator/fields/constantField",
          "validate": [Function],
        },
        "pathParams": SegmentField {
          "field": undefined,
          "key": "-ui",
          "parent": SegmentField {
            "field": undefined,
            "key": "/open-api",
            "parent": SegmentField {
              "field": undefined,
              "key": "",
              "parent": undefined,
              "type": "@spec-validator/validator/fields/segmentField",
            },
            "type": "@spec-validator/validator/fields/segmentField",
          },
          "type": "@spec-validator/validator/fields/segmentField",
        },
      },
      "response": Object {
        "body": Object {
          "getStringField": [Function],
          "regex": /\\.\\*/,
          "serialize": [Function],
          "type": "@spec-validator/validator/fields/stringField",
          "validate": [Function],
        },
        "headers": Object {
          "content-type": Object {
            "constant": "text/html",
            "getStringField": [Function],
            "regex": /text\\\\/html/,
            "serialize": [Function],
            "type": "@spec-validator/validator/fields/constantField",
            "validate": [Function],
          },
        },
        "statusCode": Object {
          "constant": 200,
          "getStringField": [Function],
          "regex": /200/,
          "serialize": [Function],
          "type": "@spec-validator/validator/fields/constantField",
          "validate": [Function],
        },
      },
    },
  ],
  "serializationFormats": Array [
    JsonSerialization {
      "deserialize": [Function],
      "mediaType": "application/json",
      "serialize": [Function],
    },
    HtmlSerialization {
      "deserialize": [Function],
      "mediaType": "text/html",
      "serialize": [Function],
    },
  ],
}
`;

exports[`routing with custom schema url 1`] = `
Object {
  "appErrorStatusCode": 500,
  "baseUrl": "http://localhost:8000",
  "encoding": "utf-8",
  "frameworkErrorStatusCode": 503,
  "reportError": [Function],
  "routes": Array [
    Object {
      "handler": [Function],
      "request": Object {
        "method": Object {
          "constant": "GET",
          "getStringField": [Function],
          "regex": /GET/,
          "serialize": [Function],
          "type": "@spec-validator/validator/fields/constantField",
          "validate": [Function],
        },
        "pathParams": SegmentField {
          "field": undefined,
          "key": "/schema",
          "parent": SegmentField {
            "field": undefined,
            "key": "",
            "parent": undefined,
            "type": "@spec-validator/validator/fields/segmentField",
          },
          "type": "@spec-validator/validator/fields/segmentField",
        },
      },
      "response": Object {
        "body": Object {
          "serialize": [Function],
          "type": "@spec-validator/validator/fields/wildcardObjectField",
          "validate": [Function],
        },
        "statusCode": Object {
          "constant": 200,
          "getStringField": [Function],
          "regex": /200/,
          "serialize": [Function],
          "type": "@spec-validator/validator/fields/constantField",
          "validate": [Function],
        },
      },
    },
    Object {
      "handler": [Function],
      "request": Object {
        "method": Object {
          "constant": "GET",
          "getStringField": [Function],
          "regex": /GET/,
          "serialize": [Function],
          "type": "@spec-validator/validator/fields/constantField",
          "validate": [Function],
        },
        "pathParams": SegmentField {
          "field": undefined,
          "key": "-ui",
          "parent": SegmentField {
            "field": undefined,
            "key": "/schema",
            "parent": SegmentField {
              "field": undefined,
              "key": "",
              "parent": undefined,
              "type": "@spec-validator/validator/fields/segmentField",
            },
            "type": "@spec-validator/validator/fields/segmentField",
          },
          "type": "@spec-validator/validator/fields/segmentField",
        },
      },
      "response": Object {
        "body": Object {
          "getStringField": [Function],
          "regex": /\\.\\*/,
          "serialize": [Function],
          "type": "@spec-validator/validator/fields/stringField",
          "validate": [Function],
        },
        "headers": Object {
          "content-type": Object {
            "constant": "text/html",
            "getStringField": [Function],
            "regex": /text\\\\/html/,
            "serialize": [Function],
            "type": "@spec-validator/validator/fields/constantField",
            "validate": [Function],
          },
        },
        "statusCode": Object {
          "constant": 200,
          "getStringField": [Function],
          "regex": /200/,
          "serialize": [Function],
          "type": "@spec-validator/validator/fields/constantField",
          "validate": [Function],
        },
      },
    },
  ],
  "serializationFormats": Array [
    JsonSerialization {
      "deserialize": [Function],
      "mediaType": "application/json",
      "serialize": [Function],
    },
    HtmlSerialization {
      "deserialize": [Function],
      "mediaType": "text/html",
      "serialize": [Function],
    },
  ],
}
`;

exports[`schema root handler 1`] = `
Object {
  "body": Object {
    "info": Object {
      "title": "Test",
      "version": "0.0.1",
    },
    "openapi": "3.0.3",
    "paths": Object {},
    "servers": Array [
      Object {
        "url": "http://localhost:8000",
      },
    ],
  },
  "statusCode": 200,
}
`;

exports[`schema root handler ui 1`] = `
Object {
  "body": "
  <html>
  <head>
    <title>Test : 0.0.1</title>

    <link rel=\\"stylesheet\\" type=\\"text/css\\" href=\\"https://unpkg.com/swagger-ui-dist@3.38.0/swagger-ui.css\\">
  <body>

    <div id=\\"swagger-ui\\"></div>

    <script src=\\"https://unpkg.com/swagger-ui-dist@3/swagger-ui-bundle.js\\" charset=\\"UTF-8\\"></script>
    <script src=\\"https://unpkg.com/swagger-ui-dist@3/swagger-ui-standalone-preset.js\\" charset=\\"UTF-8\\"></script>

    <script>
    window.onload = function() {
      const ui = SwaggerUIBundle({
        url: window.location.href.split('?')[0].replace(/-ui$/, ''),
        dom_id: '#swagger-ui',
        presets: [
          SwaggerUIBundle.presets.apis,
          SwaggerUIStandalonePreset
        ]
      })

      window.ui = ui
    }
    </script>

  </body>
  </html>
",
  "headers": Object {
    "content-type": "text/html",
  },
  "statusCode": 200,
}
`;
